JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,489 @@
+
+## File Organization and Naming
+- as it is a good practice to divide React components as container and presentational components, you should place them in separate folders (example: **containers** and **components**)
+
+```
+...
+├─ components
+├─ containers 
+...
+```
+
+- use **camelCase** for general folder naming and **PascalCase** for React components, their stylesheet files and their containing folders 
+- each component type's root folder should have an **index.js** file which exports all components, so you can later import them from a single place
+
+```
+...
+├─ components
+⎮   ├─ LoginContent
+⎮   ⎮   └─ LoginContent.js
+⎮   ⎮   └─ LoginContent.scss
+⎮   ├─ LoginHeader
+⎮   ⎮   └─ LoginHeader.js
+⎮   ⎮   └─ LoginHeader.scss
+⎮   └─ index.js
+├─ containers 
+⎮   ├─ Login
+⎮   ⎮   └─ Login.js
+⎮   ⎮   └─ Login.scss 
+⎮   └─ index.js
+...
+```
+
+## Component Declaration
+- don't use React.createClass
+- use ES2015 **class syntax** for component declaration
+
+``` javascript
+// bad
+const ComponentName = React.createClass({
+  // ...
+  render () {
+    return (
+      <div>
+        {this.componentState.message}
+      </div>
+    );
+  }
+});
+    
+// good
+import React, {Component} from 'react';
+
+class ComponentName extends Component {
+  // ...
+  render () {
+    return (
+      <div>
+        {this.componentState.message}
+      </div>
+    );
+  }
+}
+    
+export default ComponentName;
+```
+
+- if components don't have state, use **pure functions**:
+
+``` javascript
+import React from 'react';
+
+const ComponentName = ({message}) => {
+  return (
+    <div>
+      {message}
+    </div>
+  );
+};
+```
+
+## Formatting
+- use self-closing tags if the element doesn't have children
+- put **single spaces** before self-closing tags
+
+``` javascript
+// bad
+<ChildComponent onClick={this.onClickHandler}><ChildComponent/>
+}
+
+// good
+<ChildComponent onClick={this.onClickHandler} />
+```
+
+- if props don't fit on one line separate them with line breaks and also, put opening and closing tags in separate lines
+
+``` javascript
+// bad
+<ChildComponent className={styles.specialStyling}
+  title={this.title}
+  onClick={this.onClickHandler}
+/>
+
+// good
+<ChildComponent
+  className={styles.specialStyling}
+  title={this.title}
+  onClick={this.onClickHandler}
+/>
+
+// good
+<OuterChildComponent
+  className={styles.outerChild}
+  title={this.title}
+  onClick={this.onClickHandler}
+>
+  <InnerChildComponent className={styles.innerChild} />
+</OuterChildComponent>
+}
+```
+
+- always wrap *render* method's return statement in parentheses (with the exception of returning a single line)
+
+``` javascript
+// bad 
+render() {
+  // ...
+  return <div>
+      <ChildComponent />
+    </div>;
+}
+
+// bad 
+render() {
+  // ...
+  return
+    <div>
+      <ChildComponent />
+    </div>;
+}
+
+// good
+render() {
+  // ...
+  return <ChildComponent />;
```

Comment:
I like this, but

```
render (
 <ChildComponent />
);
```

scales better. What do you think?

Reply:
Added both.